main.cpp

```diff
@@ -253,481 +202,222 @@ Use a service like https://www.diffchecker.com/diff to compare your output.
-
-    void toString();
-
-private:
-    float x{0}, y{0};
-};
-
-struct A {};
-struct HeapA
-{
-    A* pointerToA = nullptr;
+    Numeric(Type t) : value( std::make_unique<Type>(t) ) { }
 
-    HeapA() : pointerToA(new A) {}
-
-    ~HeapA()
+    ~Numeric() 
     {
-        delete pointerToA;
+        value = nullptr;
     }
 
-};
-
-struct FloatType;
-struct DoubleType;
-struct IntType;
-
-struct FloatType
-{
-    FloatType(float floatPrimitive);
-    ~FloatType();
-
-    FloatType& operator+=(float f);
-    FloatType& operator-=(float f);
-    FloatType& operator*=(float f);
-    FloatType& operator/=(float f);
-
-    operator float() const { return *value; }
-
-    FloatType& pow(float f);
-    FloatType& pow(const IntType& i);
-    FloatType& pow(const FloatType& f);
-    FloatType& pow(const DoubleType& d);
-
-    /*
-        1) add two member functions named "apply()" to each of your Heap-Allocated Numeric Type wrappers.
-    */
+    operator Type() const { return *value; }
 
-    //  both apply() functions should work with chaining
-    //  2) One of the apply() functions should takes a std::function<> object as the function argument.
-    //  the std::function<> object should return *this;
-    FloatType& apply(std::function<FloatType&(float&)> func);
-
-    //  3) the other apply() function should take a function pointer. 
-    //  the function pointer should return void.
-    FloatType& apply(void(*func)(float&));
-
-private:
-    float* value = nullptr;
-    FloatType& powInternal(const float f);
-};
-
-struct DoubleType
-{
-    DoubleType(double doublePrimitive);
-    ~DoubleType();
-
-    DoubleType& operator+=(double d);
-    DoubleType& operator-=(double d);
-    DoubleType& operator*=(double d);
-    DoubleType& operator/=(double d);
+    Numeric& operator+=(Type t)
+    {
+        *value += t;
+        return *this;
+    }
 
-    operator double() const { return *value; }
+    Numeric& operator-=(Type t)
+    {
+        *value -= t;
+        return *this;
+    }
 
-    DoubleType& pow(double d);
-    DoubleType& pow(const IntType& i);
-    DoubleType& pow(const FloatType& f);
-    DoubleType& pow(const DoubleType& d);
+    Numeric& operator*=(Type t)
+    {
+        *value *= t;
+        return *this;
+    }
 
-    DoubleType& apply(std::function<DoubleType&(double&)> func);
-    DoubleType& apply(void(*func)(double&));
+    // #11
+    Numeric& operator/=(Type t)
+    {
+        // template type is an int
+        if constexpr (std::is_same<int, Type>::value)
+        {
+            // parameter's type is also an int
+            if constexpr (std::is_same<int, decltype(t)>::value)
+            {
+                // parameter is 0 don't do the division
+                if (t == 0)
+                {
+                    std::cerr << "can't divide integers by zero!" << std::endl;
+                    return *this;
+                }
+            }
+            else if ( t < std::numeric_limits<Type>::epsilon() )
```

Comment:
The flaw here is that `Type` is `int`, because you didn't template the parameter's type.
This will never be executed as long as `Type` is `int` because of the if() on 248.
- [ ] revise